src/Symfony/Component/Serializer/ArgumentResolver/UserInputResolver.php

```diff
@@ -0,0 +1,97 @@
+<?php
+
+/*
+ * This file is part of the Symfony package.
+ *
+ * (c) Fabien Potencier <fabien@symfony.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Component\Serializer\ArgumentResolver;
+
+use Symfony\Component\HttpFoundation\Request;
+use Symfony\Component\HttpKernel\Controller\ArgumentValueResolverInterface;
+use Symfony\Component\HttpKernel\ControllerMetadata\ArgumentMetadata;
+use Symfony\Component\HttpKernel\Exception\UnprocessableEntityHttpException;
+use Symfony\Component\Serializer\Annotation\Input;
+use Symfony\Component\Serializer\Exception\PartialDenormalizationException;
+use Symfony\Component\Serializer\Normalizer\DenormalizerInterface;
+use Symfony\Component\Serializer\SerializerInterface;
+use Symfony\Component\Validator\ConstraintViolation;
+use Symfony\Component\Validator\ConstraintViolationList;
+use Symfony\Component\Validator\Exception\InputValidationFailedException;
+use Symfony\Component\Validator\Validator\ValidatorInterface;
+
+/**
+ * Deserialize & validate user input.
+ *
+ * Works in duo with Symfony\Bundle\FrameworkBundle\EventListener\InputValidationFailedExceptionListener.
+ *
+ * @author Gary PEGEOT <garypegeot@gmail.com>
+ */
+class UserInputResolver implements ArgumentValueResolverInterface
+{
+    public function __construct(private SerializerInterface $serializer, private ?ValidatorInterface $validator = null)
+    {
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    public function supports(Request $request, ArgumentMetadata $argument): bool
+    {
+        return null !== $this->getAttribute($argument);
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    public function resolve(Request $request, ArgumentMetadata $argument): iterable
+    {
+        $attribute = $this->getAttribute($argument);
+        $context = array_merge($attribute->serializationContext, [
+            DenormalizerInterface::COLLECT_DENORMALIZATION_ERRORS => true,
+        ]);
+        $format = $attribute->format ?? $request->attributes->get('_format', 'json');
```

Comment:
Neither is correct, imho. `getPreferredFormat()` as well as the `_format` attribute hint at the format of the *response*. What we need here is the format of the *request payload*. This can be inferred from the `Content-Type` header.

That being said, the developer should be able to limit the set of possible formats (and content types?) via the attribute. Maybe we should even default to JSON-only, so nobody enables CSV deserialization by accident. 😓